### ISSUE TITLE
🐛Bug: The `aria-controls` attribute must point to an element in the same document

### ISSUE THREAD

**Issue by mustafa0x** [melt-ui/melt-ui#752]
https://validator.w3.org/nu/?doc=https://www.melt-ui.com/docs/builders/dialog

![image](https://github.com/melt-ui/melt-ui/assets/1473102/14766d86-370e-4b83-a739-49a18869dc13)

**Comment by TGlide** [comment]
> This is a duplicate/tightly related to #542.

I wouldn't say it is honestly. This happens because we're conditionally hiding the dialog, and maybe we shouldn't do that. Sure, we could conditionally set and unset aria-controls, but I wonder if that's okay a11y wise

**Comment by mustafa0x** [comment]
Yeah, it seems to be a different issue.

I checked react-aria and radix ui. react-aria don't have `aria-controls` on the controlling button. Radix ui does, so [validator.w3.org](https://validator.w3.org/nu/?doc=https://www.radix-ui.com/primitives/docs/components/dialog) reports the same error.

ARIA's example implementation of the Dialog (Modal) Pattern also lacks aria-controls: https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/dialog/

The concern isn't validation — it's whether pointing aria-controls to a non-existing element is a step back in a11y.

Related:
- https://heydonworks.com/article/aria-controls-is-poop/
- https://github.com/w3c/aria/issues/716 --- "That said, for some time, I have been on the verge of arguing for the deprecation of aria-controls. In every pattern where we recommend its use, I think that it is either redundant or that there is a potentially better way of establishing the relationship. But, for now that remains an open question to explore more deeply after we have solid support for all of ARIA 1.1 in all dominant assistive technology and browser combinations."

**Comment by TGlide** [comment]
> it's whether pointing aria-controls to a non-existing element is a step back in a11y.

I think we can assume it's a bad practice.

**Comment by mustafa0x** [comment]
JAWS, the only a11y tech that Heydon said supports aria-controls, seems to be rolling back support of it. Ref: https://github.com/w3c/aria/issues/995

And this open PR on aria-controls is another indication that there is no clear purpose of it. Specific to our problem here, they say "E.g., we probably don't need an explicit user exposed controlling relationship if a controlled element is [sic] directly after the controlling element in the dom". I think "created" was omitted. Ref: https://github.com/w3c/aria/pull/1996)

My takeaway: it's not worth the trouble and should just be removed for dialogs.

**Comment by mustafa0x** [comment]
On another note, role=button on <button> is superfluous, as indicated by the validator (link  in opening post).

**Comment by TGlide** [comment]
> On another note, role=button on is superfluous, as indicated by the validator (link in opening post).

We add role button in case someone is not using an actual `button` element. But again, we should probably be handling this in a smarter manner

**Comment by kelvindecosta** [comment]
This seems to be the case for the `dropdown-menu` and `menu` builders too

**Comment by huntabyte** [comment]
Feel free to open a new issue about any of the builders still experiencing this and we'll get it resolved! @kelvindecosta
